rust/template/tcp_channel/lib.rs

```diff
@@ -0,0 +1,262 @@
+    missing_docs
+)]
+
+use std::fmt::Debug;
+use std::io;
+use std::io::BufRead;
+use std::io::BufReader;
+use std::io::Write;
+use std::net::SocketAddr;
+use std::net::TcpListener;
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::ops::DerefMut;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::thread::spawn;
+use std::thread::JoinHandle;
+
+use libc::close;
+
+use observe::Observable;
+use observe::Observer;
+use observe::Subscription;
+use observe::UpdatesSubscription;
+
+use serde::de::DeserializeOwned;
+use serde::ser::Serialize;
+use serde_json::from_str;
+use serde_json::to_string;
+
+/// The receiving end of a TCP channel has an address
+/// and streams data to an observer.
+#[derive(Debug)]
+pub struct TcpReceiver<T> {
+    addr: SocketAddr,
+    listener: RawFd,
+    stream: Arc<Mutex<Option<RawFd>>>,
+    thread: JoinHandle<io::Result<()>>,
+    observer: Arc<Mutex<Option<Box<dyn Observer<T, String> + Send>>>>,
+}
+
+impl<T: DeserializeOwned + Send + Debug + 'static> TcpReceiver<T> {
+    /// Create a new TCP receiver with no observer.
+    ///
+    /// `addr` may have a port set (by setting it to 0). In such a case
+    /// the system will assign a port that is free. To retrieve this
+    /// assigned port (in the form of the full `SocketAddr`), use the
+    /// `addr` method.
+    pub fn new<A>(addr: A) -> io::Result<Self>
+    where
+        A: ToSocketAddrs,
+    {
+        let listener = TcpListener::bind(addr)?;
+        // We want to allow for auto-assigned ports, by letting the user
+        // specify a `SocketAddr` with port 0. In this case, after
+        // actually binding to an address, we need to update the port we
+        // got assigned in `addr`, but for simplicity we just copy the
+        // entire thing.
+        let addr = listener.local_addr()?;
+        let listener_fd = listener.as_raw_fd();
+        let stream_fd = Arc::new(Mutex::new(None));
+        let observer = Arc::new(Mutex::new(None));
+        let thread = Self::accept(stream_fd.clone(), listener, observer.clone());
+
+        Ok(Self {
+            addr,
+            listener: listener_fd,
+            stream: stream_fd,
+            thread,
+            observer,
+        })
+    }
+
+    /// Accept a connection (in a non-blocking manner), read data from
+    /// it, and dispatch that to the subscribed observer, if any. If no
+    /// observer is subscribed, data will be silently dropped.
+    fn accept(
+        stream: Arc<Mutex<Option<RawFd>>>,
+        listener: TcpListener,
+        observer: Arc<Mutex<Option<Box<dyn Observer<T, String> + Send>>>>,
+    ) -> JoinHandle<io::Result<()>> {
+        spawn(move || {
+            let (socket, _) = listener.accept()?;
+            *stream.lock().unwrap() = Some(socket.as_raw_fd());
+
+            let reader = BufReader::new(socket);
+            let mut lines = reader.lines();
+            loop {
```

Comment:
I don't have a better solution and this should not prevent this PR from being merged, but I wonder if there is a design that does not require `TCPReceiver` to buffer.

Reply:
What exactly is the problem with buffering? How would it be possible to ever go without buffering given that you don't know how much data you have to read in advance.

Reply:
`TCPReceiver` could theoretically feed data directly to the observer as soon as it arrives. Problem is, `TCPReceiver` can have multiple connections, all sending data at the same time. Buffering is necessary to isolate these concurrent transactions.

One option is to spawn an observable per connection and pump data to this observable, which would allow policy decisions about buffering to be pushed downstream, e.g., to whoever aggregates data from all these observables. However, this will introduce complexity, which I'd rather avoid in this early prototype.

Reply:
As we discussed in person, there is a slight misunderstanding in that a `TcpReceiver` only ever accepts a single connection (in an attempt to form a "channel"). We will always need some form of buffering in order to deserialize a message of unknown length, but we don't have to accumulate updates across `on_updates` calls. This will change shortly along with the switch of the encoder to use. Resolving.